### PR TITLE
Add an endpoint to change minion type after job created ref T64107

### PR DESCRIPTION
**Pull request by kxingit**
Add an endpoint to change minion type after job created ref T64107

### CODE REVIEW

File: hydra-main/src/main/java/com/addthis/hydra/job/web/JobRequestHandlerImpl.java [hunk 1]
```diff
@@ -116,6 +115,70 @@ public Job createOrUpdateJob(KVPairs kv, String username, String token, String s
...
+        // make a valid host set
+        Set<String> hostsWithTargetMinionType = new HashSet<String>();
+        List<HostState> hostStates = spawn.hostManager.listHostStatus(minionType);
+        for (HostState hostState : hostStates) {
+            hostsWithTargetMinionType.add(hostState.getHostUuid());
+        }
+
+        // check tasks and their replicas
+        for (JobTask jobTask : job.getCopyOfTasks()) {
+            if (!hostsWithTargetMinionType.contains(jobTask.getHostUUID())) {
```
Comment (yuesong) [inline hydra-main/src/main/java/com/addthis/hydra/job/web/JobRequestHandlerImpl.java]: Is `jobTask.getHostUUID` not one of `jobTask.getReplicas`?

File: hydra-main/src/main/java/com/addthis/hydra/job/web/resources/JobsResource.java [hunk 2]
```diff
@@ -779,7 +779,7 @@ public Response saveJob(@QueryParam("pairs") KVPairs kv,
 
     /**
      *
-     * @param jobId         job id
+     * @param job           job id
```
Comment (kxingit) [inline hydra-main/src/main/java/com/addthis/hydra/job/web/resources/JobsResource.java]: right, I was looking at `@QueryParam("job")`, that was wrong and will be fixed in the next push.

File: hydra-main/src/main/java/com/addthis/hydra/job/web/JobRequestHandlerImpl.java (hunk 1 above)
Comment (kxingit) [inline hydra-main/src/main/java/com/addthis/hydra/job/web/JobRequestHandlerImpl.java]: I read the code, but did not fully understand weather it is the case. So I ran a remote debugger for a job with `number of replicas` set to 2, in which case there are 1+2 copies in total. The debugger shows there are two replicas. I think this means `jobTask.getHostUUID` is not in `jobTask.getReplicas`. I will further read the code to confirm.

**Comment by kxingit** [comment]
Could not figure out a way to make streams/lambda more readable than a for loop. So left unchanged for now. Unit tests added, in the next commit.

File: hydra-main/src/test/java/com/addthis/hydra/job/web/JobRequestHandlerImplTest.java [hunk 3]
```diff
@@ -117,33 +117,22 @@ public void updateMinionType_notOnTargetType() throws Exception {
...
             // Expected
         }
-        verify(spawn, times(0)).updateJob(job);
-        assertSame(job.getMinionType(), "oldMinion");
+        verify(spawn, never()).updateJob(job);
+        assertEquals("oldMinion", job.getMinionType());
     }
 
-    @Test
+    @Test(expected = IllegalArgumentException.class)
```
Comment (yuesong) [inline hydra-main/src/test/java/com/addthis/hydra/job/web/JobRequestHandlerImplTest.java]: This is the convention for verifying an expected exception. I also simplified the test. I think it's a better balance between test coverage and simplicity.

**Pull request closed without merging**